### PR TITLE
fixed the transactionResult issue.

### PR DESCRIPTION
**Pull request by luodanwg**
fixed the transactionResult issue.

1.As the transactionResult func can't process with duplicate
input assetID utxo and duplicate output assetID utxo.
so divided into input result and output result func to calc
the amout group by assetid first.
2.add transfer I/O balance check.

* have test with 10w transactions and full process(reg->issue->transfer)

Signed-off-by: junjie luodan.wg@gmail.com

### CODE REVIEW

File: core/transaction/transaction.go [hunk 1]
```diff
@@ -401,28 +401,50 @@ func (tx *Transaction) GetReference() (map[*UTXOTxInput]*TxOutput, error) {
...
+		if inputValue, ok := InputResult[outputAssetid]; ok {
+			result[outputAssetid] = inputValue - outputValue
+		} else {
+			result[outputAssetid] = outputValue * Fixed64(-1)
+		}
+	}
+	return result, nil
+}
+
+func (tx *Transaction) GetMergedAssetIDValueFromOutputs() TransactionResult {
```
Comment (luodanwg) [inline core/transaction/transaction.go]: calc the output's result  group by assetid.

File: core/transaction/transaction.go [hunk 2]
```diff
@@ -401,28 +401,50 @@ func (tx *Transaction) GetReference() (map[*UTXOTxInput]*TxOutput, error) {
...
+			result[v.AssetID] = v.Value
 		}
-		if finded == false {
-			res.AssetId = o.AssetID
-			res.Amount = o.Value * Fixed64(-1)
+	}
+	return result
+}
+
+func (tx *Transaction) GetMergedAssetIDValueFromReference() (TransactionResult, error) {
```
Comment (luodanwg) [inline core/transaction/transaction.go]: calc the input's result  group by assetid.

File: core/validation/txValidator.go [hunk 3]
```diff
@@ -73,51 +73,63 @@ func VerifyTransaction(Tx *tx.Transaction, ledger *ledger.Ledger, TxPool []*tx.T
...
 			//quantity_issued : amount has been issued of this assedID
-			//txPoolAmounts   : amount in transactionPool of this assedID
-
-			// TODO: check this after the function TxStore.GetQuantityIssued works
+			//txPoolAmounts   : amount in transactionPool of this assedID of issue transaction.
 			if AssetReg.Amount-quantity_issued < txPoolAmounts {
 				return errors.New("[VerifyTransaction], Amount check error.")
 			}
 		}
+	case tx.TransferAsset:
```
Comment (luodanwg) [inline core/validation/txValidator.go]: add transferAsset input/output balance check.

**Comment by dreamfly281** [comment]
Good commit log

File: core/transaction/transaction.go [hunk 4]
```diff
@@ -310,24 +310,24 @@ func (tx *Transaction) GetProgramHashes() ([]Uint160, error) {
 		}
 		hashs = append(hashs, astHash)
 	case IssueAsset:
-		result, err := tx.GetTransactionResults()
+		result := tx.GetMergedAssetIDValueFromOutputs()
 		if err != nil {
 			return nil, NewDetailErr(err, ErrNoCode, "[Transaction], GetTransactionResults failed.")
```
Comment (dreamfly281) [inline core/transaction/transaction.go]: If there is no err return value from previous function,  why run error condition judge here?

Comment (luodanwg) [inline core/transaction/transaction.go]: Thanks for pointing out the problem. I have fixed it. could you helps to check it again,thanks.

**Pull request merged**